### PR TITLE
Make route line more configurable

### PR DESCRIPTION
**Pull request by bsudekum**
This makes the annotate not only public, but also overrideable. This way, developers can specify a custom source for route line and also styling for it too.

/cc @1ec5 @frederoni

### CODE REVIEW

**Comment by bsudekum** [comment]
Closes: https://github.com/mapbox/MapboxNavigation.swift/issues/115

File: MapboxNavigationUI/NavigationMapView.swift [hunk 1]
```diff
@@ -1,10 +1,14 @@
...
-public class NavigationMapView: MGLMapView {
+open class NavigationMapView: MGLMapView {
     
     var navigationMapDelegate: NavigationMapViewDelegate?
     
-    public override func locationManager(_ manager: CLLocationManager!, didUpdateLocations locations: [Any]!) {
+    let cap = NSValue(mglLineCap: .round)
+    let join = NSValue(mglLineJoin: .round)
+    
+    open override func locationManager(_ manager: CLLocationManager!, didUpdateLocations locations: [Any]!) {
```
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: This change appears to be unrelated. What does it address?

File: MapboxNavigationUI/NavigationMapView.swift [hunk 2]
```diff
@@ -1,10 +1,14 @@
...
+import MapboxDirections
 
-public class NavigationMapView: MGLMapView {
+open class NavigationMapView: MGLMapView {
     
     var navigationMapDelegate: NavigationMapViewDelegate?
     
-    public override func locationManager(_ manager: CLLocationManager!, didUpdateLocations locations: [Any]!) {
+    let cap = NSValue(mglLineCap: .round)
+    let join = NSValue(mglLineJoin: .round)
```
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: A boxed line join generally isn’t interesting enough to split out into a separate variable.

File: MapboxNavigationUI/NavigationMapView.swift [hunk 3]
```diff
@@ -13,4 +17,77 @@ public class NavigationMapView: MGLMapView {
...
+                    layer.identifier != arrowLayerIdentifier && layer.identifier != arrowSourceIdentifier {
+                    style.insertLayer(line, above: layer)
+                    style.insertLayer(lineCasing, below: line)
+                    return
+                }
+            }
+        }
+    }
+    
+    public func shape(for route: Route) -> MGLShape? {
```
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: s/for/describing/ to make the relationship clearer.

File: MapboxNavigationUI/NavigationMapView.swift [hunk 4]
```diff
@@ -13,4 +17,77 @@ public class NavigationMapView: MGLMapView {
             super.locationManager(manager, didUpdateLocations: locations)
         }
     }
+    
+    public func annotate(_ route: Route) {
```
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: Public APIs need public documentation.

File: MapboxNavigationUI/NavigationMapView.swift [hunk 5]
```diff
@@ -13,4 +17,77 @@ public class NavigationMapView: MGLMapView {
...
+    
+    public func shape(for route: Route) -> MGLShape? {
+        guard var coordinates = route.coordinates else {
+            return nil
+        }
+        
+        return MGLPolylineFeature(coordinates: &coordinates, count: route.coordinateCount)
+    }
+    
+    public func lineStyle(_ line: MGLLineStyleLayer) -> MGLLineStyleLayer {
```
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: This method needs a more descriptive name. Imagine reading aloud this code:

```swift
let line = lineStyle(lineLayer)
```

“Let line be line style line layer”

The reason it’s difficult to come up with a descriptive name right now is that it needs to be scoped better. This method should be responsible for creating the layer too:

```swift
public func routeStyleLayer(identifier: String, source: MGLSource) -> MGLStyleLayer
```

Now the call site could look similar to this:

```swift
let line = routeStyleLayer(identifier: "routeLayerIdentifier", source: source)
```

“Let line be the route style layer with identifier …”

File: MapboxNavigationUI/NavigationMapView.swift [hunk 6]
```diff
@@ -13,4 +17,77 @@ public class NavigationMapView: MGLMapView {
...
+        line.lineColor = MGLStyleValue(rawValue: NavigationUI.shared.tintStrokeColor.withAlphaComponent(0.6))
+        line.lineWidth = MGLStyleValue(rawValue: 5)
+        
+        line.lineCap = MGLStyleValue(rawValue: cap)
+        line.lineJoin = MGLStyleValue(rawValue: join)
+        
+        return line
+    }
+    
+    public func lineCasingStyle(_ lineCasing: MGLLineStyleLayer) -> MGLLineStyleLayer {
```
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: Similarly, this method would create the MGLLineStyleLayer: `routeCasingStyleLayer(identifier:source:)`.

File: MapboxNavigationUI/NavigationMapView.swift (hunk 1 above)
Comment (bsudekum) [inline MapboxNavigationUI/NavigationMapView.swift]: @1ec5 I think this needs to happen because of https://github.com/mapbox/mapbox-navigation-ios/pull/116/files#diff-f7f2309384eb86844377711b6c8e12caR4

Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: 👍

**Comment by bsudekum** [comment]
@1ec5 Updated, just needs documentation. Will do this once everything else is in place.

File: MapboxNavigationUI/NavigationMapView.swift [hunk 7]
```diff
@@ -35,16 +32,16 @@ open class NavigationMapView: MGLMapView {
...
             source.shape = polyline
         } else {
-            let geoJSONSource = MGLShapeSource(identifier: sourceIdentifier, shape: polyline, options: nil)
-            style.addSource(geoJSONSource)
+            let lineSource = MGLShapeSource(identifier: sourceIdentifier, shape: polyline, options: nil)
+            style.addSource(lineSource)
             
-            let line = lineStyle(MGLLineStyleLayer(identifier: routeLayerIdentifier, source: geoJSONSource))
-            let lineCasing = lineCasingStyle(MGLLineStyleLayer(identifier: routeLayerCasingIdentifier, source: geoJSONSource))
+            let line = routeStyleLayer(identifier: routeLayerIdentifier, source: lineSource)
```
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: Now that these method names are more descriptive, it might be worth adding a NavigationMapViewDelegate protocol with optional methods like `mapView(_:routeStyleLayerWithIdentifier:source:)`, and having this code call those methods. We would fall back to the `routeStyleLayer(identifier:source:)` method implemented here if those methods aren’t implemented.

This PR can land without the delegate protocol, but if subclassing turns out to be problematic, this is the solution.

File: MapboxNavigationUI/NavigationMapView.swift [hunk 8]
```diff
@@ -54,15 +57,18 @@ open class NavigationMapView: MGLMapView {
...
         guard var coordinates = route.coordinates else {
             return nil
         }
         
         return MGLPolylineFeature(coordinates: &coordinates, count: route.coordinateCount)
     }
     
-    public func routeStyleLayer(identifier: String, source: MGLSource) -> MGLStyleLayer {
+    /**
+     Function for overriding the default route line style.
```
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: These are methods, not functions. (A function stands alone, apart from any class.)

File: MapboxNavigationUI/NavigationMapView.swift [hunk 9]
```diff
@@ -54,15 +57,18 @@ open class NavigationMapView: MGLMapView {
         }
     }
     
-    public func shape(describing route: Route) -> MGLShape? {
+    open func shape(describing route: Route) -> MGLShape? {
```
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: This is publicly visible too, so it deserves some documentation.

File: MapboxNavigationUI/NavigationMapView.swift (hunk 7 above)
Comment (1ec5) [inline MapboxNavigationUI/NavigationMapView.swift]: Heh, I completely forgot that we already have a NavigationMapViewDelegate protocol that would make a nice home for these methods.
